### PR TITLE
Port remaining stream ATs as ITs

### LINKED ISSUE #3689: Port remaining stream ATs as ITs

### LINKED ISSUE #3690: Improve Task Java DSL

### LINKED ISSUE #3691: Add Task IT test coverage
Related to #3690

### PR DESCRIPTION
**Pull request by tzolov**
- Enable Prometheus docker-compose configuration.
 - Add analytics counter metrics tests for Prometheus and InfluxDB.
 - Improve the test structure.
 - Add NamedChannel tests for Destination, Tap, ManyToOne and DirectedGraph.

 Resolves #3689
 Resolves #3690
 Resolves #3691

For additional information check: https://github.com/spring-cloud/spring-cloud-dataflow/issues/3689#issuecomment-569090526

### CODE REVIEW

File: spring-cloud-dataflow-server/pom.xml [hunk 1]
```diff
@@ -53,7 +53,12 @@
 			<artifactId>spring-cloud-dataflow-rest-client</artifactId>
 			<scope>test</scope>
 		</dependency>
-
+		<dependency>
+			<groupId>net.javacrumbs.json-unit</groupId>
+			<artifactId>json-unit</artifactId>
+			<version>1.25.0</version>
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server/pom.xml]: No version in child poms

File: spring-cloud-dataflow-server/pom.xml [hunk 2]
```diff
@@ -176,7 +181,7 @@
 				<plugins>
 					<plugin>
 						<artifactId>maven-failsafe-plugin</artifactId>
-						<version>3.0.0-M3</version>
+						<version>3.0.0-M4</version>
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server/pom.xml]: These should no be defined in a global spring-cloud-dataflow-build.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposePostgresRabbitInfluxIT.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+@Suite.SuiteClasses({
+		DockerComposeTestPlatform.class,
+		DockerComposeTestStream.class,
+		DockerComposeTestTask.class
+})
+public class DockerComposePostgresRabbitInfluxIT {
+
+	private static final Logger logger = LoggerFactory.getLogger(DockerComposePostgresRabbitInfluxIT.class);
+
+	@ClassRule
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposePostgresRabbitInfluxIT.java]: We should try to use juni5 for these and see if ported rule in spring-cloud-dataflow-common-test-docker works.

File: spring-cloud-dataflow-server/pom.xml (hunk 1 above)
Comment (tzolov) [reply spring-cloud-dataflow-server/pom.xml]: right, will move it to s-c-dataflow-dependeices

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+ * @author Christian Tzolov
+ */
+public class DockerComposeFactory {
+
+	private static final Logger logger = LoggerFactory.getLogger(DockerComposeFactory.class);
+
+	/**
+	 * Data Flow version to use for the tests.
+	 */
+	public static final String DEFAULT_DATAFLOW_VERSION = "2.4.0.BUILD-SNAPSHOT";
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java]: Can we infer the default property values via a property file somehow? It feels weird to update the versions at the Java file every time we move from one version to the other.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/util/ResourceExtractor.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+					new FileOutputStream(localResourcePath.toFile()));
+			return localResourcePath.toString();
+		}
+		catch (IOException e) {
+			logger.error("Failed to extract:" + resourceUri, e);
+		}
+		return null;
+	}
+
+	public static void main(String[] args) {
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/util/ResourceExtractor.java]: Should we need this main method?

File: spring-cloud-dataflow-server/README.adoc [hunk 6]
```diff
@@ -134,16 +134,29 @@ property `spring.cloud.config.failFast=true` and the client will halt with an Ex
...
 
-. You can run only the integration tests like this:
-+
+. You can run just the integration tests like this:
+
+----
+./mvnw clean test-compile failsafe:integration-test -pl spring-cloud-dataflow-server -Pfailsafe
+----
+
+It the help of the `TestProperties` properties one can change the docker-compose files used, the exact versions of the
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-server/README.adoc]: It -> With ?

**Comment by jvalkeal** [comment]
Running `./mvnw clean install -pl spring-cloud-dataflow-server -Dtest=foo -DfailIfNoTests=false -Pfailsafe` errors:
```
[ERROR]   DockerComposeIT.analyticsCounter:593->lambda$analyticsCounter$37:597 JSON documents are different:
Array "data.result" has different length, expected: <3> but was: <0>.
Array "data.result" has different content. Missing values: [{"metric":{"__name__":"my_http_counter_total","application_guid":"${json-unit.ignore}","application_name":"counter","application_type":"sink","instance":"prometheus-rsocket-proxy:9096","instance_index":"0","job":"local-rsocket-connected","msgSize":"14","stream_name":"httpCounter"},"value":["${json-unit.ignore}","1"]}, {"metric":{"__name__":"my_http_counter_total","application_guid":"${json-unit.ignore}","application_name":"counter","application_type":"sink","instance":"prometheus-rsocket-proxy:9096","instance_index":"0","job":"local-rsocket-connected","msgSize":"29","stream_name":"httpCounter"},"value":["${json-unit.ignore}","1"]}, {"metric":{"__name__":"my_http_counter_total","application_guid":"${json-unit.ignore}","application_name":"counter","application_type":"sink","instance":"prometheus-rsocket-proxy:9096","instance_index":"0","job":"local-rsocket-connected","msgSize":"36","stream_name":"httpCounter"},"value":["${json-unit.ignore}","1"]}], expected: <[{"metric":{"__name__":"my_http_counter_total","application_guid":"${json-unit.ignore}","application_name":"counter","application_type":"sink","instance":"prometheus-rsocket-proxy:9096","instance_index":"0","job":"local-rsocket-connected","msgSize":"14","stream_name":"httpCounter"},"value":["${json-unit.ignore}","1"]},{"metric":{"__name__":"my_http_counter_total","application_guid":"${json-unit.ignore}","application_name":"counter","application_type":"sink","instance":"prometheus-rsocket-proxy:9096","instance_index":"0","job":"local-rsocket-connected","msgSize":"29","stream_name":"httpCounter"},"value":["${json-unit.ignore}","1"]},{"metric":{"__name__":"my_http_counter_total","application_guid":"${json-unit.ignore}","application_name":"counter","application_type":"sink","instance":"prometheus-rsocket-proxy:9096","instance_index":"0","job":"local-rsocket-connected","msgSize":"36","stream_name":"httpCounter"},"value":["${json-unit.ignore}","1"]}]> but was: <[]>

```

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java (hunk 4 above)
Comment (jvalkeal) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java]: What about if we start to tag these images per branch? Currently `latest` always points to master build on scdf/skipper. We could add additional tags, `latest-master`, `latest-24x`, etc when we eventually have maintenance branches.

Comment (tzolov) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java]: @ilayaperumalg , the DockerMachine and DockerComposeExtension had to be static (forgot the exact reason)! In result the standard Spring properties are not applicable.  Or do you mean something else? 
@jvalkeal,  not sure i understand.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/util/ResourceExtractor.java (hunk 5 above)
Comment (tzolov) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/util/ResourceExtractor.java]: No, i left it just as an example but will remove it

**Comment by tzolov** [comment]
@jvalkeal , maybe it is a timing issue and i need to increase the 
```
Thread.sleep(Duration.ofSeconds(30).toMillis());
```
to 60 seconds.  E.g. the time for the first micrometer metrics sent to Prometheus or Influx 

A better approach is it use the Awaitibility until the metrics are found. I will explore the second

**Comment by jvalkeal** [comment]
Ran again and error in a different test:
```
[ERROR] Tests run: 14, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 546.138 s <<< FAILURE! - in org.springframework.cloud.dataflow.integration.test.DockerComposeIT
[ERROR] namedChannelDirectedGraph  Time elapsed: 39.991 s  <<< ERROR!
org.springframework.web.client.HttpServerErrorException$InternalServerError: 500 Internal Server Error
	at org.springframework.cloud.dataflow.integration.test.DockerComposeIT.httpPost(DockerComposeIT.java:646)
	at org.springframework.cloud.dataflow.integration.test.DockerComposeIT.namedChannelDirectedGraph(DockerComposeIT.java:547)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   DockerComposeIT.namedChannelDirectedGraph:547->httpPost:646 » InternalServer 5...
[INFO] 
[ERROR] Tests run: 14, Failures: 0, Errors: 1, Skipped: 0
```

**Comment by jvalkeal** [comment]
Yeah `DockerComposeIT.analyticsCounter` is a timing issue as it didn't fail next time.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java (hunk 4 above)
Comment (tzolov) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java]: @ilayaperumalg actually you have a point. The DataFlow URL and Platform name are not used in the DockerCompose initialization but only inside the test and can be made Spring properties. I will change this

Comment (jvalkeal) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DockerComposeFactory.java]: Well I meant if we hardcode a version we could just do it for known tag?

**Comment by sabbyanandan** [comment]
This is addressed via: https://github.com/spring-cloud/spring-cloud-dataflow/commit/6504e12ea5004ebfa9bd1cb06e01f7e38952a715.

**Pull request closed without merging**